### PR TITLE
Adding unit tests

### PR DESCRIPTION
**Pull request by dmitry-shibanov**
**Description:**
In scope of this pull request we add unit tests to cover new logic.
 - [x] - Cover logic for Zulu distributor.
 - [x] - Cover logic with base class.
 - [x] - Cover logic with jdkFile
 - [x] - Cover logic for Adoptium.
 - [x] - Fix tests with authentication.

### CODE REVIEW

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    runs-on: ${{ matrix.operating-system }}
+    strategy:
+      fail-fast: false
+      matrix:
+        operating-system: [ubuntu-latest, windows-latest, macos-latest]
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Setup Node.js 12
+      uses: actions/setup-node@v1
```
Comment (maxim-lobanov) [inline .github/workflows/workflow.yml:14]: Let's revert this file before merging since I have already added it in the separate PR

File: __tests__/auth.test.ts [hunk 2]
```diff
@@ -71,7 +105,11 @@ describe('auth tests', () => {
     const password = 'TOKEN';
     const gpgPassphrase = 'GPG';
 
-    await auth.configAuthentication(id, username, password, gpgPassphrase);
+    commonObject.id = id;
+    commonObject.username = username;
+    commonObject.password = password;
+    commonObject.gpgPassphrase = gpgPassphrase;
+    await auth.configureAuthentication();
```
Comment (maxim-lobanov) [inline __tests__/auth.test.ts:112]: such kind of tests looks a bit unclear. let's move auth back to the separate method to simplify testing

File: __tests__/distributors/base.test.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  protected downloadTool(javaRelease: JavaDownloadRelease): Promise<JavaInstallerResults> {
+    throw new Error('Method not implemented.');
+  }
+
+  protected findPackageForDownload(range: semver.Range): Promise<JavaDownloadRelease> {
+    throw new Error('Method not implemented.');
+  }
+}
+
+describe('Test findInToolcache', () => {
```
Comment (maxim-lobanov) [inline __tests__/distributors/base.test.ts:28]: Use name `findInToolcache` for describe. The same for other describes

File: __tests__/distributors/base.test.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
+import * as tc from '@actions/tool-cache';
```
Comment (maxim-lobanov) [inline __tests__/distributors/base.test.ts:1]: rename file to `base-installer.test.ts`

File: __tests__/distributors/base.test.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    [
+      { version: '11.1', arch: 'x86', packageType: 'jdk' },
+      { javaVersion: actualJavaVersion, javaPath }
+    ],
+    [{ version: '11', arch: 'x86', packageType: 'jre' }, null],
+    [{ version: '8', arch: 'x86', packageType: 'jdk' }, null],
+    [
+      { version: '11.1.10', arch: 'x86', packageType: 'jdk' },
+      { javaVersion: actualJavaVersion, javaPath }
+    ],
```
Comment (maxim-lobanov) [inline __tests__/distributors/base.test.ts:68]: nitpick: sorting, put all non-null cases at top.

File: __tests__/distributors/base.test.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+  );
+
+  it.each([
+    [{ version: '11', arch: 'x86', packageType: 'jre' }],
+    [{ version: '8', arch: 'x86', packageType: 'jdk' }],
+    [{ version: '11', arch: 'x64', packageType: 'jdk' }],
+    [{ version: '11', arch: 'x64', packageType: 'jre' }]
+  ])('should throw an error for non implemented method', async input => {
+    mockJavaBase = new EmptyJavaBase(input);
+    await expect(mockJavaBase.setupJava()).rejects.toThrowError('Method not implemented.');
```
Comment (maxim-lobanov) [inline __tests__/distributors/base.test.ts:157]: I am not sure that it makes much sense to test `setupJava` method since it is pretty straightforward.
I am a bit concerned that actually we use stub class where methods not implemented and this method always throws exception. As a result, we test only begin of this method (before first call of non-implemented method).
Does it really make sense?

File: __tests__/distributors/base.test.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  it.each([
+    ['11', { version: new semver.Range('11'), stable: true }],
+    ['11.0', { version: new semver.Range('11.0'), stable: true }],
+    ['11-ea', { version: new semver.Range('11'), stable: false }],
+    ['11.0.10', { version: new semver.Range('11.0.10'), stable: true }]
+  ])('normalizeVersion from %s to %o', (input, expected) => {
+    expect(DummyJavaBase.prototype.normalizeVersion.call(null, input)).toEqual(expected);
+  });
+
+  it('normalizeVersion should throw an error for non semver', () => {
```
Comment (maxim-lobanov) [inline __tests__/distributors/base.test.ts:174]: could you please also add test case with empty line or null

File: __tests__/distributors/base.test.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+  });
+});
+
+describe('normalizeVersion', () => {
+  const DummyJavaBase = JavaBase as any;
+
+  it.each([
+    ['11', { version: new semver.Range('11'), stable: true }],
+    ['11.0', { version: new semver.Range('11.0'), stable: true }],
+    ['11-ea', { version: new semver.Range('11'), stable: false }],
```
Comment (maxim-lobanov) [inline __tests__/distributors/base.test.ts:168]: nitpick: swap 11-ea and 11.0.10 test cases
Could you please also add test case with `11.0.2-ea`

File: __tests__/distributors/local.test.ts [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+import * as tc from '@actions/tool-cache';
+import * as core from '@actions/core';
+
+import path from 'path';
+import * as semver from 'semver';
+import * as utils from '../../src/util';
+
+import { LocalDistributor } from '../../src/distributors/local/installer';
+
+describe('Test setupJava', () => {
```
Comment (maxim-lobanov) [inline __tests__/distributors/local.test.ts:12]: nitpick: describe name

File: __tests__/distributors/local.test.ts [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    jest.resetAllMocks();
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it.each([
+    [
+      { version: '8.0.289', arch: 'x64', packageType: 'jdk' },
+      expectedJdkFile,
+      { javaVersion: '8.0.289', javaPath: path.join('Java_LocalJDKFile_jdk', '8.0.289', 'x64') }
```
Comment (maxim-lobanov) [inline __tests__/distributors/local.test.ts:94]: Looks like `javaPath` uses the same value from input object. Can we build it right in test case to avoid code duplication?

File: __tests__/distributors/local.test.ts [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+      expect(utilsExtractJdkFile).toHaveBeenCalled();
+      expect(fsStat).toHaveBeenCalled();
+    }
+  });
+
+  it.each([
+    [{ version: '8.0.289', arch: 'x64', packageType: 'jdk' }, 'otherJdkFile'],
+    [{ version: '11.0.289', arch: 'x64', packageType: 'jdk' }, 'otherJdkFile'],
+    [{ version: '12.0.289', arch: 'x64', packageType: 'jdk' }, 'otherJdkFile'],
+    [{ version: '11.1.11', arch: 'x64', packageType: 'jdk' }, 'not_existing_one']
```
Comment (maxim-lobanov) [inline __tests__/distributors/local.test.ts:153]: Do we really need so many cases for exception? Looks like all of them test case where file is not found. Let's leave the single one

File: __tests__/distributors/local.test.ts [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+  });
+
+  it.each([
+    [
+      { version: '8.0.289', arch: 'x64', packageType: 'jdk' },
+      expectedJdkFile,
+      { javaVersion: '8.0.289', javaPath: path.join('Java_LocalJDKFile_jdk', '8.0.289', 'x64') }
+    ],
+    [
+      { version: '11.0.289', arch: 'x64', packageType: 'jdk' },
```
Comment (maxim-lobanov) [inline __tests__/distributors/local.test.ts:97]: I suggest reducing the number of test cases since they are pretty similar

File: __tests__/distributors/local.test.ts [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    ]
+  ])('inputs %o, jdkfile %s, result %o', async (inputs, jdkFile, expected) => {
+    mockJavaBase = new LocalDistributor(inputs, jdkFile);
+    await expect(mockJavaBase.setupJava()).resolves.toEqual(expected);
+    expect(tcFind).toHaveBeenCalled();
+    if (inputs.version === actualJavaVersion) {
+      expect(tcCacheDir).not.toHaveBeenCalled();
+      expect(utilsExtractJdkFile).not.toHaveBeenCalled();
+      expect(fsStat).not.toHaveBeenCalled();
+    } else {
```
Comment (maxim-lobanov) [inline __tests__/distributors/local.test.ts:142]: I think we need the following test cases:
1) java is resolved from toolcache, jdkfile is untouched
2) java is resolved from toolcache, jdkfile doesn't exist
3) java is unpacked from jdkfile
4) jdk file is not found

These 4 test cases are enough but it is not clear what test-cases are covered here

File: __tests__/distributors/zulu.test.ts [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+    ],
+    [
+      { version: '8-ea', arch: 'x86', packageType: 'jdk+fx' },
+      `https://api.azul.com/zulu/download/community/v1.0/bundles/?os=macos&ext=tar.gz&bundle_type=jdk&javafx=true&arch=x86&hw_bitness=32&release_status=ea&features=fx`
+    ],
+    [
+      { version: '8', arch: 'x64', packageType: 'jdk' },
+      `https://api.azul.com/zulu/download/community/v1.0/bundles/?os=macos&ext=tar.gz&bundle_type=jdk&javafx=false&arch=x86&hw_bitness=64&release_status=ga`
+    ]
+  ])('get right url for %o -> $s', async (input, parsedUrl) => {
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu.test.ts:79]: right -> correct

Comment (maxim-lobanov) [reply __tests__/distributors/zulu.test.ts:79]: `get` -> `build` or `use`

File: __tests__/distributors/zulu.test.ts [hunk 15]
```diff
@@ -0,0 +1,525 @@
...
+      { version: '8', arch: 'x86', packageType: 'jdk' },
+      `https://api.azul.com/zulu/download/community/v1.0/bundles/?os=macos&ext=tar.gz&bundle_type=jdk&javafx=false&arch=x86&hw_bitness=32&release_status=ga`
+    ],
+    [
+      { version: '8', arch: 'x86', packageType: 'jdk+fx' },
+      `https://api.azul.com/zulu/download/community/v1.0/bundles/?os=macos&ext=tar.gz&bundle_type=jdk&javafx=true&arch=x86&hw_bitness=32&release_status=ga&features=fx`
+    ],
+    [
+      { version: '8-ea', arch: 'x86', packageType: 'jdk+fx' },
+      `https://api.azul.com/zulu/download/community/v1.0/bundles/?os=macos&ext=tar.gz&bundle_type=jdk&javafx=true&arch=x86&hw_bitness=32&release_status=ea&features=fx`
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu.test.ts:73]: May be we can provide only `?os=macos&ext=tar.gz&bundle_type=jdk&javafx=true&arch=x86&hw_bitness=32&release_status=ea&features=fx
` here and concat it with `https://api.azul.com/zulu/download/community/v1.0/bundles/
` later in testcase? to improve readability

File: __tests__/distributors/zulu.test.ts [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+});
+
+describe('getArchitectureOptions', () => {
+  let zuluDistributor = new ZuluDistributor({ version: '11', arch: 'x86', packageType: 'jdk' });
+  const zuluDistributorPrototype = Object.getPrototypeOf(zuluDistributor);
+
+  it.each([
+    [{ architecture: 'x64' }, { arch: 'x86', hw_bitness: '64', abi: '' }],
+    [{ architecture: 'x86' }, { arch: 'x86', hw_bitness: '32', abi: '' }],
+    [{ architecture: 'x32' }, { arch: 'x32', hw_bitness: '', abi: '' }]
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu.test.ts:152]: It is better to use `arm` here

File: __tests__/distributors/zulu.test.ts [hunk 17]
```diff
@@ -0,0 +1,525 @@
...
+      `https://api.azul.com/zulu/download/community/v1.0/bundles/?os=macos&ext=tar.gz&bundle_type=jdk&javafx=false&arch=x86&hw_bitness=32&release_status=ga`
+    ],
+    [
+      { version: '11', arch: 'x86', packageType: 'jre' },
+      `https://api.azul.com/zulu/download/community/v1.0/bundles/?os=macos&ext=tar.gz&bundle_type=jre&javafx=false&arch=x86&hw_bitness=32&release_status=ga`
+    ],
+    [
+      { version: '11-ea', arch: 'x86', packageType: 'jdk' },
+      `https://api.azul.com/zulu/download/community/v1.0/bundles/?os=macos&ext=tar.gz&bundle_type=jdk&javafx=false&arch=x86&hw_bitness=32&release_status=ea`
+    ],
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu.test.ts:62]: Let's reduce the number of test cases:
1) x64 + jdk
2) x86 + jdk
3) x64 + jre
4) x64 + jdk+fx
5) x64 + jre+fx

Since version is not used in url, it doesn't matter

File: __tests__/distributors/zulu.test.ts [hunk 18]
```diff
@@ -0,0 +1,525 @@
...
+
+        if (url.includes('release_status=ea')) {
+          manifestData = require('../data/zulu/zulu-releases-ea.json');
+        }
+
+        if (url.includes('bundle_type=jdk')) {
+          manifestData = manifestData.filter((item: any) => item.name.includes('-jdk'));
+        } else {
+          manifestData = manifestData.filter((item: any) => item.name.includes('-jre'));
+        }
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu.test.ts:179]: Why do we need so many json files and so much logic?
What is the difference between `zulu-releases-default.json`, `zulu-releases-fx.json`, `zulu-releases-ea.json`?
I think they are almost identical so we don't need all of them

File: __tests__/distributors/zulu.test.ts [hunk 19]
```diff
@@ -0,0 +1,525 @@
...
+    ],
+    [
+      '15',
+      {
+        version: '15.0.2',
+        url: 'https://fake.cdn.azul.com/zulu/bin/zulu15.29.15-ca-jre15.0.2-macosx_x64.tar.gz'
+      },
+      { version: '15', arch: 'x86', packageType: 'jre' }
+    ]
+  ])('version is %s -> %o', async (functionInput, expected, constructorInput) => {
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu.test.ts:323]: 1) where `functionInput` is used?
2) Why do we need to compare urls? Is it not enough to compare version?

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (dmitry-shibanov) [reply .github/workflows/workflow.yml]: Done. Thank you.

File: __tests__/auth.test.ts (hunk 2 above)
Comment (dmitry-shibanov) [reply __tests__/auth.test.ts]: Done. Thank you.

File: __tests__/distributors/base.test.ts (hunk 3 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/base.test.ts]: Done. Thank you.

File: __tests__/distributors/base.test.ts (hunk 5 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/base.test.ts:66]: Done. Thank you.

File: __tests__/distributors/base.test.ts (hunk 7 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/base.test.ts:174]: Done. Thank you.

File: __tests__/distributors/base.test.ts (hunk 8 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/base.test.ts:168]: Done. Thank you.

File: __tests__/distributors/zulu.test.ts (hunk 14 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu.test.ts]: Done. Thank you.

File: __tests__/distributors/base-installer.test.ts [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+      { version: '11.1.10', arch: 'x86', packageType: 'jdk' },
+      { javaVersion: actualJavaVersion, javaPath }
+    ],
+    [{ version: '11', arch: 'x86', packageType: 'jre' }, null],
+    [{ version: '8', arch: 'x86', packageType: 'jdk' }, null],
+    [{ version: '11', arch: 'x64', packageType: 'jdk' }, null],
+    [{ version: '11', arch: 'x64', packageType: 'jre' }, null]
+  ])(`should find java for path %o -> %o`, (input, expected) => {
+    mockJavaBase = new EmptyJavaBase(input);
+    const mockJavaBasePrototype = Object.getPrototypeOf(mockJavaBase);
```
Comment (maxim-lobanov) [inline __tests__/distributors/base-installer.test.ts:84]: Let's just use `mockJavaBase['findInToolcache
']()` to invoke private method. It looks more readable

File: __tests__/distributors/base-installer.test.ts [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+
+  afterEach(() => {
+    jest.resetAllMocks();
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it.each([
+    [
+      { version: '11', arch: 'x86', packageType: 'jdk' },
```
Comment (maxim-lobanov) [inline __tests__/distributors/base-installer.test.ts:67]: nitpick: let's swap x64 and x86 in test cases. To have x64 in correct case

File: __tests__/distributors/base.test.ts (hunk 4 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/base.test.ts]: Done. Thank you.

File: __tests__/distributors/base.test.ts (hunk 6 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/base.test.ts]: Done. Thank you.

File: __tests__/distributors/local.test.ts (hunk 9 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local.test.ts]: Done. Thank you.

File: __tests__/distributors/local.test.ts (hunk 10 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local.test.ts]: Done. Thank you.

File: __tests__/distributors/local.test.ts (hunk 11 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local.test.ts:150]: Done. Thank you.

File: __tests__/distributors/local.test.ts (hunk 12 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local.test.ts]: Done. Thank you.

File: __tests__/distributors/local.test.ts (hunk 13 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu.test.ts (hunk 15 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu.test.ts (hunk 16 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu.test.ts (hunk 17 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu.test.ts (hunk 18 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu.test.ts (hunk 19 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu-installer.test.ts [hunk 22]
```diff
@@ -0,0 +1,369 @@
...
+
+describe('findPackageForDownload', () => {
+  let spyHttpClient: jest.SpyInstance;
+  let zuluDistributor: ZuluDistributor;
+
+  beforeEach(() => {
+    spyHttpClient = jest.spyOn(HttpClient.prototype, 'getJson');
+    spyHttpClient.mockImplementation(
+      async (): Promise<ifm.ITypedResponse<any[]>> => {
+        manifestData = require('../data/zulu/zulu-releases-default.json');
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu-installer.test.ts:160]: Why do we need to import this file in every `beforeEach` iteration? Can we move it to the top?

File: __tests__/distributors/zulu-installer.test.ts [hunk 23]
```diff
@@ -0,0 +1,369 @@
...
+  });
+});
+
+describe('findPackageForDownload', () => {
+  let spyHttpClient: jest.SpyInstance;
+  let zuluDistributor: ZuluDistributor;
+
+  beforeEach(() => {
+    spyHttpClient = jest.spyOn(HttpClient.prototype, 'getJson');
+    spyHttpClient.mockImplementation(
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu-installer.test.ts:158]: Use `mockReturnValue` instead of `mockImplementation`

File: __tests__/distributors/zulu-installer.test.ts [hunk 24]
```diff
@@ -0,0 +1,369 @@
...
+
+    ['8.0.282', { version: '8', arch: 'x86', packageType: 'jdk' }],
+    ['11.0.10', { version: '11', arch: 'x86', packageType: 'jdk' }],
+    ['8.0.282', { version: '8.0', arch: 'x86', packageType: 'jdk' }],
+    ['11.0.10', { version: '11.0', arch: 'x86', packageType: 'jdk' }],
+    ['8.0.282', { version: '8', arch: 'x86', packageType: 'jdk+fx' }],
+    ['11.0.10', { version: '11', arch: 'x86', packageType: 'jdk+fx' }],
+    ['15.0.2', { version: '15', arch: 'x86', packageType: 'jdk' }],
+    ['15.0.2', { version: '15-ea', arch: 'x86', packageType: 'jdk' }]
+  ])('version is %s -> %o', async (expected, constructorInput) => {
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu-installer.test.ts:189]: swap input and expected parameters.

File: __tests__/distributors/zulu-installer.test.ts [hunk 25]
```diff
@@ -0,0 +1,369 @@
...
+
+  it('Should throw an error', async () => {
+    zuluDistributor = new ZuluDistributor({ version: '18', arch: 'x86', packageType: 'jdk' });
+    await expect(
+      zuluDistributor['findPackageForDownload'](zuluDistributor['version'])
+    ).rejects.toThrowError(/Could not find satisfied version for semver */);
+  });
+});
+
+describe('setupJava', () => {
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu-installer.test.ts:204]: Let's get rid of setupJava tests since we have already tested them in base-installer

File: __tests__/distributors/zulu-installer.test.ts [hunk 26]
```diff
@@ -0,0 +1,369 @@
...
+          result: JSON.parse(result) as IZuluVersions[]
+        };
+      }
+    );
+
+    originalPlatform = process.platform;
+    Object.defineProperty(process, 'platform', {
+      value: 'darwin'
+    });
+  });
```
Comment (maxim-lobanov) [inline __tests__/distributors/zulu-installer.test.ts:39]: Why do we need to override process.platform? Can be just mock `getPlatformOption`?

Comment (dmitry-shibanov) [reply __tests__/distributors/zulu-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu-installer.test.ts (hunk 22 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu-installer.test.ts]: Removed. Thank you.

File: __tests__/distributors/zulu-installer.test.ts (hunk 23 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu-installer.test.ts]: Done .Thank you.

File: __tests__/distributors/zulu-installer.test.ts (hunk 24 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu-installer.test.ts (hunk 25 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/adoptium-installer.test.ts [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+        const expectedUrl = `${baseUrl}?project=jdk&vendor=adoptopenjdk&heap_size=normal&jvm_impl=hotspot&sort_method=DEFAULT&sort_order=DESC&${expectedParameters}`;
+        distributor['getPlatformOption'] = () => 'mac';
+        await distributor['getAvailableVersions']();
+        expect(spyHttpClient.mock.calls).toHaveLength(1);
+        expect(spyHttpClient.mock.calls[0][0]).toBe(expectedUrl);
+      }
+    );
+  });
+
+  afterEach(() => {
```
Comment (AlenaSviridenko) [inline __tests__/distributors/adoptium-installer.test.ts:60]: Is it intended to have this afterEach in the root of describe? it will run after the whole `describe('findPackageForDownload')` and not after each test in that suite

File: __tests__/distributors/base-installer.test.ts [hunk 28]
```diff
@@ -0,0 +1,210 @@
...
+    ],
+    [
+      { version: '11', arch: 'x64', packageType: 'jdk' },
+      { javaPath: `/toolcache/Java_Empty_jdk/11.0.8`, javaVersion: '11.0.8' }
+    ],
+    [
+      { version: '11', arch: 'x64', packageType: 'jre' },
+      { javaPath: `/toolcache/Java_Empty_jre/11.0.8`, javaVersion: '11.0.8' }
+    ]
+  ])('download java', async (input, expected) => {
```
Comment (AlenaSviridenko) [inline __tests__/distributors/base-installer.test.ts:169]: please, add parameters to test description, otherwise it is not obvious what exact configuration failed when all 3 tests are called "download java"

File: __tests__/distributors/base-installer.test.ts [hunk 29]
```diff
@@ -0,0 +1,210 @@
...
+    expect(spyTcFind).toHaveBeenCalled();
+    expect(spyCoreAddPath).toHaveBeenCalled();
+    expect(spyCoreExportVariable).toHaveBeenCalled();
+    expect(spyCoreSetOutput).toHaveBeenCalled();
+  });
+
+  it.each([
+    [{ version: '15', arch: 'x86', packageType: 'jre' }],
+    [{ version: '11.0.7', arch: 'x64', packageType: 'jre' }]
+  ])('should throw an error for Available version not found', async input => {
```
Comment (AlenaSviridenko) [inline __tests__/distributors/base-installer.test.ts:181]: same here

File: __tests__/distributors/local-installer.test.ts [hunk 30]
```diff
@@ -0,0 +1,194 @@
...
+    });
+
+    spyTcCacheDir = jest.spyOn(tc, 'cacheDir');
+    spyTcCacheDir.mockImplementation(
+      (archivePath: string, toolcacheFolderName: string, version: string, architecture: string) =>
+        path.join(toolcacheFolderName, version, architecture)
+    );
+
+    // Spy on core methods
+
```
Comment (AlenaSviridenko) [inline __tests__/distributors/local-installer.test.ts:50]: extra empty line

File: __tests__/distributors/local-installer.test.ts [hunk 31]
```diff
@@ -0,0 +1,194 @@
...
+    expected.javaPath = path.join('Java_LocalJDKFile_jdk', inputs.version, inputs.arch);
+    await expect(mockJavaBase.setupJava()).resolves.toEqual(expected);
+    expect(spyTcFind).toHaveBeenCalled();
+    expect(spyCoreInfo).toHaveBeenCalledWith(`Resolved Java ${actualJavaVersion} from tool-cache`);
+    expect(spyCoreInfo).not.toHaveBeenCalledWith(
+      `Java ${inputs.version} is not found in tool-cache. Trying to unpack JDK file...`
+    );
+  });
+
+  it("java is resolved from toolcache, jdkfile doesn't exist", async () => {
```
Comment (AlenaSviridenko) [inline __tests__/distributors/local-installer.test.ts:108]: different quotes

File: __tests__/distributors/local-installer.test.ts [hunk 32]
```diff
@@ -0,0 +1,194 @@
...
+      `Java ${inputs.version} is not found in tool-cache. Trying to unpack JDK file...`
+    );
+  });
+
+  it.each([
+    [{ version: '8.0.289', arch: 'x64', packageType: 'jdk' }, 'otherJdkFile'],
+    [{ version: '11.0.289', arch: 'x64', packageType: 'jdk' }, 'otherJdkFile'],
+    [{ version: '12.0.289', arch: 'x64', packageType: 'jdk' }, 'otherJdkFile'],
+    [{ version: '11.1.11', arch: 'x64', packageType: 'jdk' }, 'not_existing_one']
+  ])('inputs %s, jdkfile %s', async (inputs, jdkFile) => {
```
Comment (AlenaSviridenko) [inline __tests__/distributors/local-installer.test.ts:175]: test name is unclear, what is expected, what do you check in this test?

File: __tests__/distributors/local-installer.test.ts [hunk 33]
```diff
@@ -0,0 +1,194 @@
...
+
+  it.each([
+    [{ version: '8.0.289', arch: 'x64', packageType: 'jdk' }, ''],
+    [{ version: '13.0.289', arch: 'x64', packageType: 'jdk' }, ''],
+    [{ version: '16.0.289', arch: 'x64', packageType: 'jdk' }, ''],
+    [{ version: '7.0.289', arch: 'x64', packageType: 'jdk' }, undefined],
+    [{ version: '11.0.289', arch: 'x64', packageType: 'jdk' }, undefined],
+    [{ version: '12.0.289', arch: 'x64', packageType: 'jdk' }, undefined],
+    [{ version: '15.0.289', arch: 'x64', packageType: 'jdk' }, undefined]
+  ])('inputs %s, jdkfile %s', async (inputs, jdkFile) => {
```
Comment (AlenaSviridenko) [inline __tests__/distributors/local-installer.test.ts:189]: same here

File: __tests__/distributors/zulu-installer.test.ts [hunk 34]
```diff
@@ -0,0 +1,139 @@
...
+    ],
+    [
+      { version: '8', arch: 'x64', packageType: 'jdk+fx' },
+      '?os=macos&ext=tar.gz&bundle_type=jdk&javafx=true&arch=x86&hw_bitness=64&release_status=ga&features=fx'
+    ],
+    [
+      { version: '8', arch: 'x64', packageType: 'jre+fx' },
+      '?os=macos&ext=tar.gz&bundle_type=jre&javafx=true&arch=x86&hw_bitness=64&release_status=ga&features=fx'
+    ]
+  ])('build correct url for %s -> $s', async (input, parsedUrl) => {
```
Comment (AlenaSviridenko) [inline __tests__/distributors/zulu-installer.test.ts:57]: $s is not resolved in tests output

File: __tests__/distributors/zulu-installer.test.ts [hunk 35]
```diff
@@ -0,0 +1,139 @@
...
+    zuluDistributor = new ZuluDistributor(input);
+    zuluDistributor['getPlatformOption'] = () => 'macos';
+    const buildUrl = `https://api.azul.com/zulu/download/community/v1.0/bundles/${parsedUrl}`;
+    await zuluDistributor['getAvailableVersions']();
+    expect(spyHttpClient.mock.calls).toHaveLength(1);
+    expect(spyHttpClient.mock.calls[0][0]).toBe(buildUrl);
```
Comment (AlenaSviridenko) [inline __tests__/distributors/zulu-installer.test.ts:63]: please, split setup/action/expect in tests by new lines for easier reading

File: src/check.ts [hunk 36]
```diff
@@ -0,0 +1,5 @@
+import semver from 'semver';
```
Comment (AlenaSviridenko) [inline src/check.ts:1]: what is the purpose of this file?

File: __tests__/distributors/adoptium-installer.test.ts [hunk 37]
```diff
@@ -0,0 +1,138 @@
...
+    jest.resetAllMocks();
+    jest.clearAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it('load available versions', async () => {
```
Comment (AlenaSviridenko) [inline __tests__/distributors/adoptium-installer.test.ts:66]: I think this test can be moved to getAvailableVersions suite, like you did for zulu tests

Comment (dmitry-shibanov) [reply __tests__/distributors/adoptium-installer.test.ts:68]: I think we should keep it in internal describe because the outer is `getAvailableVersions`

File: src/check.ts (hunk 36 above)
Comment (dmitry-shibanov) [reply src/check.ts]: Done. Thank you.

File: __tests__/distributors/zulu-installer.test.ts (hunk 35 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/zulu-installer.test.ts (hunk 34 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/zulu-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/local-installer.test.ts (hunk 33 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/local-installer.test.ts (hunk 32 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/local-installer.test.ts (hunk 31 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local-installer.test.ts:108]: Done. Thank you.

File: __tests__/distributors/base-installer.test.ts (hunk 29 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/base-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/base-installer.test.ts (hunk 28 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/base-installer.test.ts]: Done. Thank you.

File: __tests__/distributors/adoptium-installer.test.ts (hunk 27 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/adoptium-installer.test.ts:62]: We have internal afterEach and outer.

File: __tests__/distributors/local-installer.test.ts (hunk 30 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/local-installer.test.ts]: Done. Thank you

File: __tests__/distributors/adoptium-installer.test.ts (hunk 37 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/adoptium-installer.test.ts:60]: Done. Thank you

File: __tests__/distributors/adoptium-installer.test.ts (hunk 27 above)
Comment (dmitry-shibanov) [reply __tests__/distributors/adoptium-installer.test.ts:22]: Done. Thank you

**Review by AlenaSviridenko — APPROVED** [review]

**Review by maxim-lobanov — APPROVED** [review]
